src/bentoml/_internal/gradio.py

```diff
@@ -0,0 +1,71 @@
+from __future__ import annotations
+
+import typing as t
+from pathlib import Path
+
+from bentoml.exceptions import MissingDependencyException
+
+R = t.TypeVar("R")
+
+try:
+    import gradio as gr
+    from gradio import Blocks
+except ImportError:  # pragma: no cover
+    raise MissingDependencyException(
+        """'gradio' is required by 'bentoml.gradio', run 'pip install -U gradio' to install gradio.""",
+    )
+
+
+def mount_gradio_app(
+    blocks: Blocks,
+    path: str,
+    name: str = "gradio_ui",
+):
+    from _bentoml_sdk.service import Service
+    from bentoml._internal import server
+
+    favicon_path = (
+        Path(server.__file__).parent / "static_content" / "favicon-light-32x32.png"
+    )
+    assert path.startswith("/"), "Routed paths must start with '/'"
+    path = path.rstrip("/")
+
+    def decorator(obj: R) -> R:
+        blocks.dev_mode = False
+        blocks.show_error = True
+        blocks.validate_queue_settings()
+        blocks.root_path = path
+        blocks.favicon_path = favicon_path
+        gradio_app = gr.routes.App.create_app(blocks, app_kwargs={"root_path": path})
+
+        # @bentoml.service() decorator returns a wrapper instead of the original class
+        # Check if the object is an instance of Service
+        if isinstance(obj, Service):
+            # For scenario:
+            #
+            # @bentoml.gradio.mount_gradio_app(..)
+            # @bentoml.service(..)
+            # class MyService: ...
+            #
+            # If the Service instance is already created, mount the ASGI app immediately
+            target = obj.inner
+            obj.mount_asgi_app(gradio_app, path=path, name=name)
```

Comment:
It seems not necessary to mount it here, otherwise it will be mounted twice?

Reply:
it has to do with the Service wrapper class creation time, similar to the mount_asgi_app method here: https://github.com/bentoml/BentoML/blob/main/src/_bentoml_sdk/decorators.py#L124